src/FlashLoanVault.sol

```diff
@@ -0,0 +1,285 @@
+        burn(owner, shares);
+        
+        //transfer assets
+        underlying.safeTransfer(receiver, assets);
+        emit Withdraw(msg.sender, receiver, owner, assets, shares);
+    }
+
+    /// @dev Burns shares from owner and sends assets of underlying tokens to receiver; based on the exchange rate.
+    /// @param shares The amount of wrapped tokens to redeem for underlying tokens (assets)
+    /// @param receiver Address of receiver of underlying tokens - DAI
+    /// @param owner Address of owner of Fractional Wrapper shares - yvDAI
+    function redeem(uint256 shares, address receiver, address owner) external returns(uint256 assets) {
+        assets = convertToAssets(shares);
+        
+        // MUST support a redeem flow where the shares are burned from owner directly where owner is msg.sender,
+        // OR msg.sender has ERC-20 approval over the shares of owner
+        if(msg.sender != owner){
+            uint256 allowedShares = _allowance[owner][receiver] ;
+            require(allowedShares >= shares, "Allowance exceeded!");
+            _allowance[owner][receiver] = allowedShares - shares;
+        }
+
+        //burn wrapped tokens(shares) -> yvDAI 
+        burn(owner, shares);
+        
+        //transfer assets
+        underlying.safeTransfer(receiver, assets);
+        emit Withdraw(msg.sender, receiver, owner, assets, shares);
+    }
+
+
+
+    /// @notice Returns the unit 'exchange rate'; assuming 1 unit of underlying was deposited, how much shares would be received 
+    /// @dev wrapperMinted = (underlyingDeposited,(1) * wrapperSupply) / underlyingInWrapper 
+    /// Note: Exchange rate is floating, it's dynamic based on capital in/out-flows
+    /// Note: _totalSupply returns a value extended over decimal precision, in this case 18 dp. hence the scaling before divsion.
+    function getExchangeRate() internal view returns(uint256) {
+        uint256 sharesMinted;
+        return _totalSupply == 0 ? sharesMinted = 1e18 : sharesMinted = (_totalSupply * 1e18) / underlying.balanceOf(address(this));
+    }
+
+    /// @notice Calculate how much yvDAI user should get based on flaoting exchange rate
+    /// @dev getExchangeRate() returns shares minted per unit of underlying asset deposited; at present moment.
+    /// @param assets Amount of underlying tokens (assets) to be converted to wrapped tokens (shares)
+    function convertToShares(uint256 assets) internal view returns(uint256 shares){
+        return assets * getExchangeRate() / 1e18;
+    }
+
+    /// @notice Calculate how much DAI user should get based on floating exchange rate
+    /// @dev getExchangeRate() returns shares minted per unit of underlying asset deposited; at present moment.
+    /// @param shares Amount of wrapped tokens (shares) to be converted to underlying tokens (assets) 
+    function convertToAssets(uint256 shares) internal view returns(uint256 assets){
+        return shares * 1e18 / getExchangeRate();
```

Comment:
As above:
`return _totalSupply == 0 ? 1e18 : shares * underlying.balanceOf(address(this)) / _totalSupply;`

Reply:
Updated as well:

```solidity
    function convertToAssets(uint256 shares) internal view returns(uint256 assets){
        return _totalSupply == 0 ? shares : shares * underlying.balanceOf(address(this)) / _totalSupply;
    }
````